### ISSUE TITLE
Modeling DSLs

### ISSUE THREAD

**Issue by cscherrer** [cscherrer/Soss.jl#100]
Soss has a lot of aspects that aren't specific to its modeling approach, or even to Bayesian analysis at all:

- Map a user-supplied `@model` AST to a `Model` type
- Include in `Model` type:
    - Module where the model was defined
    - Type-level representation of the model
- Code generation:
    - Take a mapping from a `Model` type and other arguments to an AST
    - Use this to define a new method, taking care of tracking the module and the type-level representation of the model

---

This approach can clearly be used in other contexts. This raises some questions:

- Should Soss be expressed as an instance of some more general model-building library?
- Is this approach even more general than "modeling"? Maybe it should be a "DSL builder" library?

I'm leaning toward thinking the first is a "yes". We could conceivably even arrive at a single front end that allows Gen, Turing, and Soss implementations as back ends!

**Comment by cscherrer** [comment]
Very generally, we can consider `@model` as using environment and argument information to initialize a struct, and then folding the lines of the body into that struct. This fold could do any computation we like as it goes, or it could just put the lines back into an Expr as it began with. 

This approach has a lot in common with the "free monad" approach sometimes used in the Haskell community, but all of this happens at compile time

**Comment by ChrisRackauckas** [comment]
>Is this approach even more general than "modeling"? Maybe it should be a "DSL builder" library?

And now you've arrived at https://github.com/SciML/ModelingToolkit.jl . It's two things, first a CAS which is faster and (soon to be) more expansive than SymPy. But secondly, it's a model definition and transformation system built on top of said CAS. We cover linear, nonlinear, discrete dynamic, (stochastic) ordinary differential equation, control, reaction, partial differential equation, and optimization systems right now. It would be interesting to discuss what the internal format would be for a probabilistic system, or whether that is captured by one of the current AST formats + support for distributions.

**Comment by cscherrer** [comment]
MeasureTheory.jl should be a big help there, Distributions.jl has a fair amount of overhead and wasn't really built for this sort of thing. I had some really nice accelerations using Sympy, it would be great to port those ideas over to ModelingToolkit as well. It would be really interesting if it could work as a back-end for Soss.

**Comment by ChrisRackauckas** [comment]
Yeah, what kinds of accelerations do you implement? We've been writing a lot of passes recently to do things like delete redundant variables and do log-transforms, those kinds of things. I am curious what a PPL uses.

**Comment by cscherrer** [comment]
The one that got the big speedups goes something like this:

For a given distribution, the logpdf is typically a sum of a few different terms. For a multivariate value, it's just a sum of these. 

Say we have this sum written as a `for` loop. Then the body of that loop will have some terms that depend on the loop index, and some that don't. Using this to distribute the sum can make a big difference.

CAS can get us there, but we should also be able to do this for exponential families without any symbolic work (or I guess really it's that the symbolic stuff would be implicit)
